### PR TITLE
Add `Pkg.freeable` to list packages that might be freed without downgrading

### PR DESCRIPTION
**Pull request by timholy**
There's some incentive to run tagged releases as much as possible: it makes `Pkg.update` a little faster, and it helps ensure that the developer of the package sees the same bugs that most users do. So, freeing your packages periodically is a good idea, but this can be hard to remember. I've been using this little utility off-and-on for quite a long time, and I decided it might be time to polish it up and see if folks think it's worth supporting in general. (If it doesn't belong in base, maybe it could go to PkgDev.)

Demo:

``` jl
julia> f = freeable()
Packages with a gap between HEAD and the most recent tag:
ASTInterpreter: 7
BinDeps: 14
Compat: 27
DWARF: 5
ELF: 3
Gallium: 16
Gtk: 28
Images: 6
JuliaParser: 1
ObjFileBase: 2
Reactive: 3
TerminalUI: 3
VT100: 2
8-element Array{Any,1}:
 "AbstractTrees"   
 "ColorVectorSpace"
 "FactCheck"       
 "HDF5"            
 "ImageMagick"     
 "MachO"           
 "SnoopCompile"    
 "StructIO"

julia> Pkg.free(f)
INFO: Freeing AbstractTrees
INFO: Freeing ColorVectorSpace   
...
```

The text output at the top can help you realize that some packages have seen a lot of development without being tagged.

CC @wildart; I am skeptical that I'm really using `Pkg` the way it should be used, so critique from you would be gratefully accepted. (This was basically minimally ported from something I developed in the 0.3 or 0.4 timeframe.)

### CODE REVIEW

**Comment by tkelman** [comment]
This would be better for PkgDev IMO.

**Comment by IainNZ** [comment]
Its a nice idea, though!

File: base/pkg/entry.jl [hunk 1]
```diff
@@ -290,6 +290,52 @@ function free(pkgs)
...
+        instd = Read.installed()
+        first = true
+        k, v = collect(keys(instd)), collect(values(instd))
+        p = sortperm(k)
+        k = k[p]
+        v = v[p]
+        for (pkg,status) in zip(k,v)
+            ver, fix = status
+            if fix
+                repo = GitRepo(pkg)
```
Comment (wildart) [inline base/pkg/entry.jl]: use `LibGit2.with` function to invoke `do` pattern to guaranty resource deallocation. 
```julia
LibGit2.with(GitRepo(pkg)) do repo
# working with `repo`
end
```

File: base/pkg/entry.jl [hunk 2]
```diff
@@ -290,6 +290,52 @@ function free(pkgs)
...
+                head = LibGit2.head(pkg)
```
Comment (wildart) [inline base/pkg/entry.jl]: `LibGit2.head_oid(repo)` to get sha1 identifier

File: base/pkg/entry.jl [hunk 3]
```diff
@@ -290,6 +290,52 @@ function free(pkgs)
...
+                sha == nothing && continue
+                tagged = sha.sha1
+                if head != tagged
+                    if first
+                        println(io, "Packages with a gap between HEAD and the most recent tag:")
+                        first = false
+                    end
+                    n = -1
+                    try
+                        vrs = LibGit2.revcount(GitRepo(pkg), tagged, "HEAD")
```
Comment (wildart) [inline base/pkg/entry.jl]: `LibGit2.revcount(repo, tagged, head)`

**Comment by timholy** [comment]
Thanks for the helpful tips, @wildart!

Moved to https://github.com/JuliaLang/PkgDev.jl/pull/42

**Comment by kmsquire** [comment]
I'd actually like to see a `Pkg.pinned()` function (which seems like a simpler version of this PR.  I don't see that as a `PkgDev` function, since pinning has uses outside of package development.

**Pull request closed without merging**